spring-integration-core/src/main/java/org/springframework/integration/config/SmartLifecycleRoleController.java

```diff
@@ -0,0 +1,194 @@
+		this.lifecycles.add(role, lifecycle);
+	}
+
+	/**
+	 * Add a {@link SmartLifecycle} bean to the role using its name.
+	 * @param role the role.
+	 * @param lifecycleBeanName the bean name of the {@link SmartLifecycle}.
+	 */
+	public void addLifecycleToRole(String role, String lifecycleBeanName) {
+		Assert.state(this.applicationContext != null, "An application context is required to use this method");
+		this.lazyLifecycles.add(role, lifecycleBeanName);
+	}
+
+	/**
+	 * Add a {@link SmartLifecycle} beans to the role using their names.
+	 * @param role the role.
+	 * @param lifecycleBeanNames the bean names of the {@link SmartLifecycle}s.
+	 */
+	public void addLifecyclesToRole(String role, List<String> lifecycleBeanNames) {
+		Assert.state(this.applicationContext != null, "An application context is required to use this method");
+		for (String lifecycleBeanName : lifecycleBeanNames) {
+			this.lazyLifecycles.add(role, lifecycleBeanName);
+		}
+	}
+
+	/**
+	 * Start all registered {@link SmartLifecycle}s in the role.
+	 * @param role the role.
+	 */
+	public void startLifecyclesInRole(String role) {
+		if (this.lazyLifecycles.size() > 0) {
+			addLazyLifecycles();
+		}
+		List<SmartLifecycle> lifecycles = this.lifecycles.get(role);
+		if (lifecycles != null) {
+			for (SmartLifecycle lifecycle : lifecycles) {
+				try {
+					lifecycle.start();
+				}
+				catch (Exception e) {
+					logger.error("Failed to start " + lifecycle + " in role " + role);
+				}
+			}
+		}
+	}
+
+	/**
+	 * Stop all registered {@link SmartLifecycle}s in the role.
+	 * @param role the role.
+	 */
+	public void stopLifecyclesInRole(String role) {
+		if (this.lazyLifecycles.size() > 0) {
+			addLazyLifecycles();
+		}
+		List<SmartLifecycle> lifecycles = this.lifecycles.get(role);
+		if (lifecycles != null) {
+			lifecycles = new ArrayList<SmartLifecycle>(lifecycles);
+			Collections.reverse(lifecycles);
+			for (SmartLifecycle lifecycle : lifecycles) {
+				try {
+					lifecycle.stop();
+				}
+				catch (Exception e) {
+					logger.error("Failed to stop " + lifecycle + " in role " + role);
+				}
+			}
+		}
+	}
+
+	private void addLazyLifecycles() {
+		for (Entry<String, List<String>> entry : this.lazyLifecycles.entrySet()) {
+			doAddLifecyclesToRole(entry.getKey(), entry.getValue());
+		}
+		this.lazyLifecycles.clear();
+	}
+
+	private void doAddLifecyclesToRole(String role, List<String> lifecycleBeanNames) {
+		for (String lifecycleBeanName : lifecycleBeanNames) {
+			try {
+				SmartLifecycle lifecycle = this.applicationContext.getBean(lifecycleBeanName, SmartLifecycle.class);
+				addLifecycleToRole(role, lifecycle);
+			}
+			catch (NoSuchBeanDefinitionException e) {
+				logger.warn("Skipped; no such bean :" + lifecycleBeanName);
+			}
+		}
+	}
+
+	@Override
+	public void onApplicationEvent(AbstractLeaderEvent event) {
+		if (event instanceof OnGrantedEvent) {
+			startLifecyclesInRole(event.getRole());
+		}
+		else if (event instanceof OnRevokedEvent) {
```

Comment:
Looks like these are only a reason for `s-c-c-c` dependency. So, we should rework it to the class names and reflection for the getter.
Otherwise I'm sure that dependency can be `optional` and this `SmartLifecycleRoleController` can be registered only if the first one is in CP.